File: solutions/3920-minimum-stability-factor-of-array/minimum-stability-factor-of-array.java
class Solution {
    static int[] tree;
    static int n;

   
    static void build(int[] nums) {
        n = nums.length;
        tree = new int[4 * n]; 
        buildUtil(nums, 0, 0, n - 1);
    }

    static void buildUtil(int[] nums, int node, int start, int end) {
        if (start == end) {
            tree[node] = nums[start];
        } else {
            int mid = (start + end) / 2;
            buildUtil(nums, 2 * node + 1, start, mid);
            buildUtil(nums, 2 * node + 2, mid + 1, end);
            tree[node] = gcd(tree[2 * node + 1], tree[2 * node + 2]);
        }
    }

  
    static int query(int l, int r) {
        return queryUtil(0, 0, n - 1, l, r);
    }
    
    static int queryUtil(int node, int start, int end, int l, int r) {
        if (r < start || end < l) {
            return 0;
        }
        
        if (l <= start && end <= r) {
            return tree[node];
        }
        int mid = (start + end) / 2;
        int leftGCD = queryUtil(2 * node + 1, start, mid, l, r);
        int rightGCD = queryUtil(2 * node + 2, mid + 1, end, l, r);
        return gcd(leftGCD, rightGCD);
    }
	
	
	public int minStable(int[] nums, int maxC) {

        if(maxC == nums.length) return 0;

        int lo = 1;
        int hi = nums.length;
        int ans = nums.length;
        boolean allOne = true;
        int countOfNotOne = 0;
        tree = new int[8*nums.length];
        
        for(int i = 0;i<nums.length;i++){
            if(nums[i]!=1){
                allOne = false;
                countOfNotOne++;
            }
        }
        if(allOne || countOfNotOne ==maxC){
            return 0;
        }
        build(nums);
        while(lo<=hi){
            int mid = (lo+hi)/2;

            boolean ok = isPossible(nums, maxC,mid);

            if(ok){
                ans = mid;
                hi = mid-1;
            }else{
                lo = mid+1;
            }
        }

        return ans;
        
    }

    static boolean isPossible(int []nums, int max, int mid){

        int gcd = 0;
        int i = 0;
        
        while(i<nums.length) {
        	
        	if((i+mid)>=nums.length) break;
        	
        	gcd = query(i,i+mid);
        	
        	
        	if(gcd >=2) {
        		i = i+mid;
        		max--;
        	}
        	i++;
        }
        
        
        
        
        return (max>=0);
    }

    static int gcd(int a, int b){
        if(b==0) return a;

        return gcd(b, a%b);
    }
    
}

File: solutions/3920-minimum-stability-factor-of-array/minimum-stability-factor-of-array.cpp
class GCDSparseTable {
private:
    vector<vector<int>> st;
    vector<int> logVal;
    int n;

public:
    GCDSparseTable(const vector<int>& a) {
        n = a.size();
        int maxLog = 32 - __builtin_clz(n);
        st.assign(n, vector<int>(maxLog));
        logVal.assign(n + 1, 0);

        for (int i = 2; i <= n; i++)
            logVal[i] = logVal[i / 2] + 1;

        for (int i = 0; i < n; i++)
            st[i][0] = a[i];

        for (int j = 1; (1 << j) <= n; j++) {
            for (int i = 0; i + (1 << j) <= n; i++) {
                st[i][j] = __gcd(st[i][j - 1], st[i + (1 << (j - 1))][j - 1]);
            }
        }
    }

    int getGCD(int l, int r) {
        int j = logVal[r - l + 1];
        return __gcd(st[l][j], st[r - (1 << j) + 1][j]);
    }
};

class Solution {
public:
    vector<int> a;
    int n, maxC;

    bool canBreak(int maxLen, GCDSparseTable& st) {
        int changes = 0, l = 0;

        for (int r = 0; r < n; r++) {
            int g = st.getGCD(l, r);

            while (l + 1 <= r && g == 1) {
                l++;
                g = st.getGCD(l, r);
            }

            if (r - l + 1 > maxLen) {
                l = r + 1;
                changes++;
            }
        }
        return changes <= maxC;
    }

    int minStable(vector<int>& nums, int _maxC) {
        a = nums;
        n = a.size();
        maxC = _maxC;

        if (count(a.begin(), a.end(), 1) + maxC >= n) return 0;

        GCDSparseTable st(a);
        int l = 0, r = n + 1;

        while (r > l + 1) {
            int mid = (l + r) / 2;
            if (canBreak(mid, st)) r = mid;
            else l = mid;
        }
        return r;
    }
};